src/handlers/VehicleTestController.ts

```diff
@@ -1,58 +1,230 @@
+   */
+  private async generateExpiryDate(
+    payload: models.ITestResultPayload
+  ): Promise<models.ITestResultPayload> {
+    try {
+      if (
+        payload.testStatus !== enums.TEST_STATUS.SUBMITTED ||
+        utils.ValidationUtil.isNotAllowedVehicleTypeForExpiry(
+          payload.vehicleType
+        )
+      ) {
+        return Promise.resolve(payload);
+      }
+      const expiryTestTypes = payload.testTypes.filter((testType) =>
+        utils.ValidationUtil.isAllowedTestTypeForExpiry(testType)
+      );
+
+      const recentExpiry = await this.dataProvider.getMostRecentExpiryDate(
+        payload.systemNumber
+      );
+
+      expiryTestTypes.forEach((testType: any, index: number) => {
+        const testTypeForExpiry: models.TestTypeForExpiry = {
+          testType,
+          vehicleType:
+            enums.VEHICLE_TYPE[
+              payload.vehicleType.toUpperCase() as keyof typeof enums.VEHICLE_TYPE
+            ],
+          recentExpiry,
+          regnOrFirstUseDate: VehicleTestController.getRegistrationOrFirstUseDate(payload),
+          hasHistory: !DateProvider.isSameAsEpoc(recentExpiry),
+          hasRegistration: DateProvider.isValidDate(
+            VehicleTestController.getRegistrationOrFirstUseDate(payload)
+          ),
+        };
+        console.log(testTypeForExpiry);
+        const strategy = this.getExpiryStrategy(testTypeForExpiry);
+        console.log(strategy.constructor.name);
+        testType.testExpiryDate = strategy.getExpiryDate();
+      });
+      console.log("generateExpiryDate: testTypes ->", payload.testTypes);
+      return Promise.resolve(payload);
+    } catch (error) {
+      console.error("Error in error generateExpiryDate", error);
+      throw new models.HTTPError(500, enums.MESSAGES.INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  private static AssignCertificateNumberToTestTypes(
+    payload: models.ITestResultPayload
+  ) {
+    if (payload.testStatus !== enums.TEST_STATUS.SUBMITTED) {
+      return payload;
+    }
+    const { vehicleType } = payload;
+    payload.testTypes
+      .filter(
+        (testType) =>
+          utils.ValidationUtil.isPassedRoadworthinessTestForHgvTrl(
+            vehicleType,
+            testType.testTypeId,
+            testType.testResult
+          ) ||
+          (utils.ValidationUtil.isAnnualTestTypeClassificationWithoutAbandonedResult(
+            testType.testTypeClassification,
+            testType.testResult
+          ) &&
+            !utils.ValidationUtil.isTestTypeAdr(testType) &&
+            !utils.ValidationUtil.isTestTypeLec(testType))
+      )
+      .forEach((testType) => {
+        testType.certificateNumber = testType.testNumber;
+      });
+    return payload;
+  }
+
+  private static calculateAnniversaryDate(payload: models.ITestResultPayload) {
+    const { vehicleType } = payload;
+    payload.testTypes.forEach((testType: models.TestType) => {
+      const { testExpiryDate } = testType;
+      if (!testExpiryDate) {
```

Comment:
I will need to catch up on that one please, I need to udnerstand the requirements better I think